### PR TITLE
AP_Logger: make LogStructure non-packed to fix compilation of SITL on Apple M1

### PR DESCRIPTION
**Pull request by ntamas**
As of ArduCopter 4.2.0-rc2, the compilation of the SITL simulator fails on Apple M1 with `clang` (version 13.1.6); the compiler emits the following error messages during the linking phase:

```
ld: warning: pointer not aligned at address 0x100363D61 (__ZN6Copter13log_structureE + 7129 from ArduCopter/Log.cpp.44.o)
ld: warning: pointer not aligned at address 0x100363D59 (__ZN6Copter13log_structureE + 7121 from ArduCopter/Log.cpp.44.o)
ld: warning: pointer not aligned at address 0x100363D51 (__ZN6Copter13log_structureE + 7113 from ArduCopter/Log.cpp.44.o)
ld: warning: pointer not aligned at address 0x100363D49 (__ZN6Copter13log_structureE + 7105 from ArduCopter/Log.cpp.44.o)
[... lots of similar messages ...]
ld: warning: pointer not aligned at address 0x10036219A (__ZN6Copter13log_structureE + 18 from ArduCopter/Log.cpp.44.o)
ld: warning: pointer not aligned at address 0x100362192 (__ZN6Copter13log_structureE + 10 from ArduCopter/Log.cpp.44.o)
ld: warning: pointer not aligned at address 0x10036218A (__ZN6Copter13log_structureE + 2 from ArduCopter/Log.cpp.44.o)
ld: unaligned pointer(s) for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Removing `PACKED` from `struct LogStructure` fixes the problem. A short discussion with @andyp1per on Discord seems to indicate that this particular structure does not need to be packed as it does not represent the exact byte-level layout of a concrete log record that we may write to the SD card or a disk.

Note that you need a Mac with an Apple M1 CPU to reproduce the problem. Compiling the SITL for an Intel CPU does not exhibit this behaviour; even on an Apple M1, cross-compiling for Intel works correctly. The problem appears only when targeting an Apple M1 CPU during SITL compilation.

### CODE REVIEW

**Comment by timtuxworth** [comment]
Thanks for working on this, I appreciate the fix.

**Review by andyp1per — APPROVED** [review]

**Review by WickedShell — CHANGES_REQUESTED** [review]
This increases flash space a fair amount as it raises the size per LogStructure on a 32 bit arm target from 23 bytes per, to 28 bytes per. If we really need to do this then we should consider this at a minimum:

``` c++
// structure used to define logging format    
struct LogStructure {    
    const char *name;    
    const char *format;    
    const char *labels;    
    const char *units;    
    const char *multipliers;    
    uint8_t msg_type;                                                                                                                 
    uint8_t msg_len;    
    bool streaming; // can be rate limited    
}; 
```

Rearranging like this means this only costs 1 more byte of flash per log structure, and per copy.

I haven't actually checked that we don't emit the packed structure directly, but the change as presented eats a lot more memory then we'd like it to.

**Comment by IamPete1** [comment]
```
Binary Name      Text [B]        Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
 ---------------  --------------  -----------  -------------  ----------------------------  -------------------------
ardusub          752 (+0.0484%)  0 (0.0000%)  0 (0.0000%)    752 (+0.0484%)                                   410732
arducopter       772 (+0.0438%)  0 (0.0000%)  -4 (-0.0015%)  772 (+0.0437%)                                   197620
blimp            796 (+0.0616%)  0 (0.0000%)  -4 (-0.0015%)  796 (+0.0615%)                                   671216
ardurover        736 (+0.0463%)  0 (0.0000%)  0 (0.0000%)    736 (+0.0462%)                                   372736
arducopter-heli  776 (+0.0438%)  0 (0.0000%)  0 (0.0000%)    776 (+0.0437%)                                   190044
arduplane        796 (+0.0456%)  0 (0.0000%)  4 (+0.0015%)   796 (+0.0455%)                                   216864
antennatracker   728 (+0.0539%)  0 (0.0000%)  0 (0.0000%)    728 (+0.0538%)                                   612800
```

**Comment by IamPete1** [comment]
Maybe we can `#ifdef` this for Apple M1 only? I don't think we want to pay the flash cost across the board.

**Comment by timtuxworth** [comment]
> Maybe we can `#ifdef` this for Apple M1 only? I don't think we want to pay the flash cost across the board.

Or just for SITL?

**Comment by ntamas** [comment]
Good point. I can do either, but it seems like narrowing down to Apple M1 with preprocessor macros is not that easy. The best I can come up with is `#if defined(APPLE) && defined(__aarch64__)` -- would that be okay?

**Comment by andyp1per** [comment]
TBH just predicating it on SITL seems fine to me

**Comment by ntamas** [comment]
Modified the PR so the change affects SITL only.

**Comment by timtuxworth** [comment]
It seems that the major impact of using uint8_t for the message type and message number is to limit the message number to < 256. If 16 bits were used  for these two fields, and therefore no need for packed, (yes a small cost - a few hundred bytes) - you could have 16,000 message numbers. Surely that would be worth it?

**Comment by timtuxworth** [comment]
Also the only other field impacted by PACKED is msg_len also uint8_t, but invariably set to 50 or 51 (using hard coded values!) everywhere in the code that it is used. I bet a lot more flash could be saved by using a global constant for this value, removing this field from the structure and therefore removing all the places in the code where the msg_len is calculated, passed, processed and stored, but effectively not used. If this was done, 16 bits could be used for the type with a net saving due to all the code that could be refactored out.

File: libraries/AP_Logger/LogStructure.h [hunk 1]
```diff
@@ -136,7 +136,13 @@ const struct MultiplierStructure log_Multipliers[] = {
 #include <AP_AIS/LogStructure.h>
 
 // structure used to define logging format
+// It is packed to save flash space; however, this causes problems when building
+// the SITL on an Apple M1 CPU so we do not pack it when compiling SITL
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
```
Comment (tridge) [inline libraries/AP_Logger/LogStructure.h:141]: make it PACKED only on HAL_CHIBIOS
that will ensure all the linux ports are correct, and flash space is minimized on stm32

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_Logger/LogStructure.h (hunk 1 above)
Comment (ntamas) [reply libraries/AP_Logger/LogStructure.h]: done

**Review by tridge — APPROVED** [review]

**Review by WickedShell — APPROVED** [review]

**Comment by tridge** [comment]
squashed commits and pushed to master
thanks!

**Pull request closed without merging**